sdk/core/Azure.Core/src/ResourceIdentifier.cs

```diff
@@ -0,0 +1,429 @@
+                Location = parent.Location;
+                ResourceGroupName = parent.ResourceGroupName;
+            }
+
+            if (resourceType == SubscriptionResourceType)
+            {
+                if (!Guid.TryParse(resourceName, out _))
+                    throw new ArgumentOutOfRangeException(nameof(resourceName), $"The GUID for subscription is invalid {resourceName}.");
+                SubscriptionId = resourceName;
+            }
+
+            if (resourceType.GetLastType() == LocationsKey)
+                Location = resourceName;
+
+            if (resourceType == ResourceGroupResourceType)
+                ResourceGroupName = resourceName;
+
+            if (resourceType == ProviderResourceType)
+                Provider = resourceName;
+        }
+
+        private static ResourceType ChooseResourceType(string resourceTypeName, ResourceIdentifier parent)
+        {
+            if (resourceTypeName.Equals(ResourceGroupsLowerKey, StringComparison.OrdinalIgnoreCase))
+                return ResourceGroupResourceType;
+
+            //subscriptions' type is Microsoft.Resources/subscriptions only when its parent is Tenant
+            if (resourceTypeName.Equals(SubscriptionsKey, StringComparison.OrdinalIgnoreCase) && parent.ResourceType == TenantResourceType)
+                return SubscriptionResourceType;
+
+            return parent.ResourceType.AppendChild(resourceTypeName);
+        }
+
+        private static ResourceIdentifier Create(string resourceId)
+        {
+            Argument.AssertNotNullOrEmpty(resourceId, nameof(resourceId));
+
+            if (resourceId[0] != Separator)
+                throw new ArgumentOutOfRangeException(nameof(resourceId), "The ResourceIdentifier must start with '/'.");
+
+            if (resourceId.Length == 1)
+                return Root;
+
+            string[] parts = resourceId.Split(Separator);
+            if (parts.Length < 3)
+                throw new ArgumentOutOfRangeException(nameof(resourceId), "The ResourceIdentifier is too short it must have at least 2 path segments.");
+
+            if (!parts[1].Equals(SubscriptionsKey, StringComparison.OrdinalIgnoreCase) && !parts[1].Equals(ProvidersKey, StringComparison.OrdinalIgnoreCase))
+                throw new ArgumentOutOfRangeException(nameof(resourceId), $"The ResourceIdentifier must start with either '/{SubscriptionsKey}' or '/{ProvidersKey}'.");
+
+            int validLength = string.IsNullOrEmpty(parts[parts.Length - 1]) ? parts.Length - 1 : parts.Length;
+            return AppendNext(Root, parts, 1, validLength);
+        }
+
+        private static ResourceIdentifier AppendNext(ResourceIdentifier parent, string[] parts, int startIndex, int length)
+        {
+            int partsCount = length - startIndex;
+
+            if (partsCount == 0)
+                return parent;
+
+            string firstElement = parts[startIndex];
+
+            if (partsCount == 1)
```

Comment:
As we talked about, comments on what we create in each of these sections would help maintainers here.